Improve Lambda sample application

trevorbonas opened the pull request:
*Issue #, if available:*

N/A.

*Description of changes:*

- Majority of the Python code has been put into its own module.
- The Python code has been divided up into multiple functions.
- The sample application directory has been restructured, with a `notebooks/` directory containing `demo.ipynb` and `src/lambda_sample/` containing the Python code for creating the Lambda function, creating the Amazon Managed Grafana workspace, and generating time series data.
- Each data generator now generates its own Grafana dashboard, customized to fit its data.
- The permissions the Lambda function uses to write to Timestream for LiveAnalytics is now more specific.
- Instead of being a string, the code for the Lambda function is now in `lambda_function.py`.
- Retries with exponential backoffs are now used when creating the Lambda's IAM role and sending data to the Lambda function, as the Lambda function can take a while to become operational.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

trevorbonas committed:
Improve Lambda sample application

*Issue #, if available:*

N/A.

*Description of changes:*

- Majority of the Python code has been put into its own module.
- The Python code has been divided up into multiple functions.
- The sample application directory has been restructured, with a `notebooks/` directory containing `demo.ipynb` and `src/lambda_sample/` containing the Python code for creating the Lambda function, creating the Amazon Managed Grafana workspace, and generating time series data.
- Each data generator now generates its own Grafana dashboard, customized to fit its data.
- The permissions the Lambda function uses to write to Timestream for LiveAnalytics is now more specific.
- Instead of being a string, the code for the Lambda function is now in `lambda_function.py`.
- Retries with exponential backoffs are now used when creating the Lambda's IAM role and sending data to the Lambda function, as the Lambda function can take a while to become operational.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

sethusrinivasan commented on integrations/lambda/src/lambda_sample/lambda_function.py:
```diff
@@ -0,0 +1,140 @@
+                prepared_records.append(prepared_record)
+
+            # Write to Timestream using the `write_records` API
+            response = timestream_client.write_records(
```
Would be better to have an alternate strategy to ingest records via BatchLoad (creating csv files, copy the generated file and data model  to S3 and create BatchLoad task ) - https://docs.aws.amazon.com/timestream/latest/developerguide/code-samples.create-batch-load.html

trevorbonas replied:
Users can now either:
1. Have the Lambda function ingest data directly into Timestream for LiveAnalytics.
2. Have the Lambda function create a batch load task that ingests data into Timestream for LiveAnalytics.

By default, option 1 is used.

For option 2, users must provide the names of the S3 buckets they've created themselves to use to hold data and hold any errors logs during ingestion. These names are sent to the Lambda function as query strings. When a task is successfully created, the user will be given a 202 status code, indicating that their data is being ingested.

sethusrinivasan commented on integrations/lambda/src/lambda_sample/lambda_function.py:
```diff
@@ -0,0 +1,140 @@
+        else:
+            raise e
+
+def lambda_handler(event, context):
```
Post ingestion, can this lambda ingestor also run a Timestream Query to report back results of few queries like
1/ Table metadata -> "DESCRIBE db.table" https://docs.aws.amazon.com/timestream/latest/developerguide/supported-sql-constructs.DESCRIBE.html
2/ few statistics - SELECT MIN(time), MAX(time), count(*) from db.table

trevorbonas replied:
Both of these queries are now executed after ingesting records. The results of the queries can be viewed in the Lambda function's logs.

sethusrinivasan commented on integrations/lambda/src/lambda_sample/lambda_function.py:
```diff
@@ -0,0 +1,140 @@
+
+    try:
+        # Create table if it does not exist
+        timestream_client.create_table(
```
Let us also make sure that a customer defined partition key is define on newly created table - https://docs.aws.amazon.com/timestream/latest/developerguide/customer-defined-partition-keys.html

trevorbonas replied:
Customers can now define a partition key to use for newly-created Timestream for LiveAnalytics tables. It will default to `'MEASURE'` unless users specify a dimension to use.

sethusrinivasan commented on integrations/lambda/src/lambda_sample/lambda_helper.py:
```diff
@@ -0,0 +1,296 @@
+
+def send_request(session: session, method: str, function_url: str, payload: dict, precision="MILLISECONDS"):
+    """
+    Sends a single request to the Lambda function.
```
can comment state what kind of request with an example?

trevorbonas replied:
I have added an example here of how the payload should be formatted.

sethusrinivasan commented on integrations/lambda/src/lambda_sample/lambda_helper.py:
```diff
@@ -0,0 +1,296 @@
+MAX_WAIT_SECONDS = 900 # 15 minutes
+
+def create_lambda(session: session, lambda_name: str, database_name: str, table_name: str, role_name: str,
+                  mem_store_retention_period_in_hours=12, mag_store_retention_period_in_days=3653, batch_size=100) -> str:
```
can we align these defaults with TImestream console's "Create Table"'s defaults?

trevorbonas replied:
The default values for `mem_store_retention_period_in_hours` and `mag_store_retention_period_in_days` match the default values in the Timestream console, 12 hours and 10 years respectively.

sethusrinivasan commented on integrations/lambda/src/lambda_sample/lambda_function.py:
```diff
@@ -0,0 +1,140 @@
+    """Create Timestream database and table if they do not exist."""
+    try:
+        # Create database if it does not exist
+        timestream_client.create_database(DatabaseName=DATABASE_NAME)
```
create a database with customer managed key?

trevorbonas replied:
Customers can now use a KMS key ID of their choosing.

sethusrinivasan commented on integrations/lambda/src/lambda_sample/data_generator/air_quality_data_generator.py:
```diff
@@ -0,0 +1,93 @@
+from .data_generator import DataGenerator
+
+class AirQualityDataGenerator(DataGenerator):
```
Can you also add unit test cases along with this change?

trevorbonas replied:
I have added unit tests for all data generators.

sethusrinivasan commented on integrations/lambda/README.md:
```diff
@@ -21,6 +21,7 @@ The following diagram depicts the deployed Lambda function receiving generated d
 ## Prerequisites
```
This is a data generation utility and wondering why it is names "lambda" integration?. pls rename the project folder and move it under /tools

trevorbonas replied:
I have changed the name of the application to "Timestream for LiveAnalytics Sample Data Generator," since it generates samples of real-life time series data. The generator is now in `tools/python/sample-data-generator/`.

trevorbonas committed:
Move data generator to tools/python/sample-data-generator

trevorbonas committed:
Rename lambda_sample package

trevorbonas committed:
Add customer-defined partition key support

trevorbonas committed:
Add support for KMS key ID when creating database

trevorbonas committed:
Query table after ingestion is complete

trevorbonas committed:
Add example of request

trevorbonas committed:
Add the option to use a batch load task for ingestion

fredjoonpark committed:
Improve flight paths for the `FlightDataGenerator` class

Issue #, if available:

N/A.

Description of changes:

- `FlightDataGenerator` now generates more realistic flight paths, using airport coordinates around the world to spherically interpolate data points using the great-circle distance formula.

- Added a new "direction" field to the `DataGenerator` class, which species the allowed direction of change for the measure value over time. I added it to track fuel levels (which we dont ever expect to go up mid-flight). @trevorbonas let me know if this new field feels weird to you or if you think I should keep this within the custom subclass.

- Added unit tests (inferring we'll need it gauging by the upstream review comments)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

trevorbonas committed:
Add unit tests for all data generators

Signed-off-by: Trevor Bonas <Trevor.Bonas@improving.com>

trevorbonas committed:
Remove unused constant in lambda_function
